File: Java/Factory.java
package Java;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

/**
 * 多线程任务分发器
 * 自动将任务分配给当前任务最少的 Worker 执行。
 * 与 JavaScript 版 Factory 功能对应。
 * 基于 jinyaoMa 的代码修改而来，增加了 Worker 的概念，使得任务分发更加合理。
 * 
 * @author SuiMu
 */

public class Factory<T> {

  public interface Listener<T> {

    void onComplete(Factory<T> factory);

  }

  public static class Task<T> {

    private static long nextId = 0;

    public interface Process<T> {

      void run(Task<T> task, Next<T> next);

      void onResolve(T data);

      void onReject(Exception e);

    }

    public interface Next<T> {

      void resolve(T data);

      void reject(Exception e);
    }

    private long id;

    private Process<T> process;

    private Next<T> next = new Next<T>() {

      @Override
      public void resolve(T data) {
        process.onResolve(data);
      }

      @Override
      public void reject(Exception e) {
        process.onReject(e);
      }
    };

    public Task(Process<T> process) {
      id = nextId++;
      setProcess(process);
    }

    public long getId() {
      return id;
    }

    public void setProcess(Process<T> process) {
      this.process = process;
    }

    public void run() {
      try {
        process.run(this, next);
      } catch (Exception e) {
        process.onReject(e);
      }
    }
  }

  public static class Worker<T> {

    private static long nextId = 0;

    private long id;
    private Queue<Task<T>> tasks;
    private Thread thread = null;
    private Runnable runnable = new Runnable() {

      @Override
      public void run() {
        while (!tasks.isEmpty()) {
          Task<T> task = tasks.poll();
          task.run();
        }
      }
    };

    public Worker() {
      id = nextId++;
      tasks = new LinkedList<Task<T>>();
    }

    public long getId() {
      return id;
    }

    public int getTasksCount() {
      return tasks.size();
    }

    public boolean isWorking() {
      return thread != null && thread.isAlive();
    }

    public void add(Task<T> task) {
      tasks.add(task);
      if (!isWorking()) {
        thread = new Thread(runnable);
        thread.start();
      }
    }
  }

  private final int MIN_WORKER_SIZE = 1; // minimum size - 1 worker
  private ArrayList<Worker<T>> workers;

  public Factory(int size) throws IllegalArgumentException {
    if (size < MIN_WORKER_SIZE) {
      throw new IllegalArgumentException("Minimum size is 1");
    }
    workers = new ArrayList<Worker<T>>(size);
    for (int i = 0; i < size; i++) {
      workers.add(new Worker<T>());
    }
  }

  public Worker<T> getWorkerWithTheLeastTasks() {
    int size = workers.size();
    if (size > 0) {
      int min = Integer.MAX_VALUE;
      int index = 0;
      for (int i = 0; i < size; i++) {
        int count = workers.get(i).getTasksCount();
        if (count < min) {
          min = count;
          index = i;
        }
      }
      return workers.get(index);
    }
    return null;
  }

  public void add(Task<T> task) {
    Worker<T> workerWithTheLeastTasks = getWorkerWithTheLeastTasks();
    workerWithTheLeastTasks.add(task);
  }

  public void add(ArrayList<Task<T>> tasks) {
    for (Task<T> task : tasks) {
      add(task);
    }
  }

  public boolean isEmpty() {
    boolean result = true;
    for (Worker<T> worker : workers) {
      if (worker.isWorking() || worker.getTasksCount() > 0) {
        result = false;
        break;
      }
    }
    return result;
  }
}

File: JavaScript/factory.js
/**
 * Factory - 异步任务队列管理器
 * 按顺序执行任务，支持停止和重启。
 * 基于 jinyaoMa 的代码修改
 * 
 * @author SuiMu
 */

class Factory {
  constructor() {
    this.queue = [];
    this.isWorking = false;
    this.forceStop = false;
  }
  add(task) {
    if (task instanceof Factory.Task) {
      this.queue.push(task);
      this.run();
    } else {
      throw new Error('Factory.add: invalid arguments');
    }
  }
  run() {
    this.forceStop = false;
    if (!this.isWorking) {
      this.next();
    }
  }
  next() {
    if (this.queue.length) {
      let task = this.queue.shift();
      task.on(Factory.Task.START, o => {
        this.isWorking = true;
      }).on(Factory.Task.COMPLETE, result => {
        if (result !== undefined && result !== null && !this.forceStop) {
          task.listeners && typeof task.listeners.oncomplete === 'function' && task.listeners.oncomplete(result);
          this.next();
        } else {
          this.isWorking = false;
          this.forceStop = false;
        }
      }).on(Factory.Task.ERROR, error => {
        this.isWorking = false;
        this.forceStop = false;
        task.listeners && typeof task.listeners.onerror === 'function' && task.listeners.onerror(error);
        console.error(error);
      }).do();
    }
  }
  stop() {
    this.forceStop = true;
  }
}
Factory.Task = class {
  constructor(callback, listeners = null) {
    if (typeof callback === 'function') {
      this.callback = callback;
    } else {
      throw new Error('Factory.Tack.constructor: invalid arguments');
    }
    if (listeners != null) {
      this.listeners = listeners;
    }
  }
  on(action, callback) {
    if (typeof action === 'string' && typeof callback === 'function') {
      switch (action.toLowerCase().trim()) {
        case Factory.Task.START:
          this.onstart = callback;
          break;
        case Factory.Task.COMPLETE:
          this.oncomplete = callback;
          break;
        case Factory.Task.ERROR:
          this.onerror = callback;
          break;
      }
    } else {
      throw new Error('Factory.Tack.on: invalid arguments');
    }
    return this;
  }
  do() {
    this.onstart && this.onstart();
    this.callback(this.oncomplete, this.onerror);
  }
}
Factory.Task.START = 'start';
Factory.Task.COMPLETE = 'complete';
Factory.Task.ERROR = 'error';